data_diff/diff_tables.py

```diff
@@ -239,12 +245,36 @@ def _diff_tables(self, table1, table2, level=0, segment_index=None, segment_coun
         for res in self._thread_map(list, diff_iters):
             yield from res
 
+    def _diff_tables(self, table1, table2, level=0, segment_index=None, segment_count=None):
+        logger.info(
+            ". " * level
+            + f"Diffing segment {segment_index}/{segment_count}, "
+            f"key-range: {table1.start_key}..{table2.end_key}, "
+            f"size: {table2.end_key-table1.start_key}"
+        )
+
+        (count1, checksum1), (count2, checksum2) = self._threaded_call("count_and_checksum", [table1, table2])
+
+        if count1 == 0 and count2 == 0:
+            logger.warn(
+                "Uneven distribution of keys detected. (big gaps in the key column). "
+                "For better performance, we recommend to increase the bisection-threshold."
+            )
+            assert checksum1 is None and checksum2 is None
+            return
+
+        if level == 1:
+            self.stats["table1_count"] = self.stats.get("table1_count", 0) + count1
```

Comment:
Wouldn't it make more sense to set this on the `TableSegment`? `table1` is kind of an internal vocabulary, this would be a bit hard to support as an external API. It's OK with me if you don't want to worry about this right now, and want to address that with the Python API, and just get it merged.

I think this goes to your idea of a parallelized count, too. To me a "parallelized count" would be calling `diff_tables`, but just not doing `Checksum()` in the query—only the count. That could be a cool idea 🤷🏻

Reply:
We want `TableSegment` to be a reusable structure. For example, to be used by two diffs at the same time. If we accumulate the sum over there, we'll have two threads in a race condition that is almost certain to happen.

So I think it's right to accumulate inside of `TableDiffer`. We can, afterwards, store the result in TableSegment. This is something that's easy to add later too.

A parallelized count would use the same base mechanism as diff_tables, but it would run a different sequence, since it's only 1 table, so a lot of the diff operations like checksums and comparisons aren't necessary. It makes sense for it to be in a different function.

Anyway, I think both of these, if we want them, belong in a new PR.